runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java

```diff
@@ -19,22 +19,67 @@
 package org.apache.beam.runners.spark.coders;
 
 import com.esotericsoftware.kryo.Kryo;
+import com.google.common.base.Function;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
 import de.javakaffee.kryoserializers.UnmodifiableCollectionsSerializer;
 import de.javakaffee.kryoserializers.guava.ImmutableListSerializer;
 import de.javakaffee.kryoserializers.guava.ImmutableMapSerializer;
 import de.javakaffee.kryoserializers.guava.ImmutableMultimapSerializer;
 import de.javakaffee.kryoserializers.guava.ImmutableSetSerializer;
 import de.javakaffee.kryoserializers.guava.ReverseListSerializer;
+import java.util.Arrays;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.io.Source;
 import org.apache.spark.serializer.KryoRegistrator;
+import org.reflections.Reflections;
 
 
 /**
  * Custom {@link com.esotericsoftware.kryo.Serializer}s for Beam's Spark runner needs.
  */
 public class BeamSparkRunnerRegistrator implements KryoRegistrator {
 
+  private static final Class<?>[] CLASSES_FOR_JAVA_SERIALIZATION = new Class<?>[]{
+      Coder.class, Source.class
+  };
+
+  private static volatile Iterable<Class<?>> subclassesForJavaSerialization;
+
+  /**
+   * Register coders and sources with {@link StatelessJavaSerializer} since sources and coders
+   * aren't guaranteed to be Kryo-serializable.
+   **/
+  private static Iterable<Class<?>> getClassesForJavaSerialization() {
```

Comment:
Double lock + volatile should be fine here, but you could also consider "init on demand holder" pattern.